### PR TITLE
docs: report on storage changes and decisions. 

### LINKED ISSUE #1044: Benchmark current storage improvements
We need to create a testing infrastructure focused on storage in order to test and validate the storage optimizations and changes we plan to implement based on the outcome of the issues where we define metrics and understand workloads (#46),  create a baseline (#67), and evaluate database engines (#64).

In order to create this testing environment, the following tasks should be performed:

- Identify a workload that generates enough data on storage for meaningful testing (at least 10s of GB) on a single machine. 
- Containerized (Docker) environment for testing that can be run locally and in a CI environment
- Support for multiple ComeBFT binaries compiled with all the supported databases (e.g. LevelDB, RocksDB)
- Allow testing for multiple database configuration options
- Generate testing output data that can be used to generate reports that can display improvements (or not) against a baseline or any other particular branch/commit.

17.01.2024.

The scope of the issue has been changed and here is the plan on what will be completed during Q1 2024. 
- [x] Identify a workload that generates enough data for meaningful storage.
- [x] Record the size of the databases during the experiment

## Definition of Done (DoD):

- [x] Decided and fixed workload that will be used to benchmark any storage improvements. 
- [x] Running the workload with metrics done in #46 on baseline CometBFT without the optimizations
- [x] Running the workload with metrics done in #46 on CometBFT with optimizations from #1041 and #49. 
- [ ] Produce a report comparing the results obtained by all the experiments.  #2596

## **Deprioritized - left for later (open issue to continue tracking if needed):**
- Any end user (e.g. developer) can run this testing infrastructure locally without too much local configuration and can generate testing data for validating the database optimizations during the development process
- The storage testing e2e runs in a CI infrastructure (Github actions) (optional, it is not yet well defined and not clear what it would contain). 
- Generated reports that can be shared with the community that outline the improvements accomplished as a result of optimizations and new implementations

### PR DESCRIPTION
**Pull request by jmalicevic**
[Rendered](https://github.com/cometbft/cometbft/tree/jasmina/storage_report/docs/references/storage)

Closes #1044 

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

**Comment by jmalicevic** [comment]
I wonder whether in the report we can include a link to the  branch that does migration from v1 to v2 to incentivise users to use it: https://github.com/cometbft/cometbft/blob/migrate_db/cmd/cometbft/commands/migrate_db.go . I am not sure it should be merged into main though.

**Review by melekes — APPROVED** [review]
love this ❤️

**Comment by melekes** [comment]
> I wonder whether in the report we can include a link to the branch that does migration from v1 to v2 to incentivise users to use it: https://github.com/cometbft/cometbft/blob/migrate_db/cmd/cometbft/commands/migrate_db.go 

I think it should be in the `experimental_key_layout` docs.

File: docs/references/storage/README.md [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+The one experiment with a real application did not back up this theory though. Even though the overall performance was better than the software version used by the application at the moment. 
+
+The block processing time reported by our application were in the range of 2-6s compared to 100s of ms for the real world application. Furthermore, the application might have different access patterns. Our tests can be seen as testing only CometBFT's interaction with storage, without much interference from the application.  
+
+That is why, in Q1, we introduce an interface with two implementations: the current key layout and a new representation, sorting the keys by height using ordercode. The new layout is marked as purely experimental. Our hope is that chains will be incentivized to experiment with it and provide us with more real world data. This will also facilitate switching the data layout without breaking changes between releases if we decide to officially support a new data layout. 
+
```
Comment (adizere) [inline docs/references/storage/README.md:39]:
```suggestion

**Summary of Q1 results:**

- pruning on its own is inefficient to control storage growth
- we have complemented pruning with a (forced) compaction feature, and this proved effective to control storage growth
- to mitigate potential performance that pruning + compaction may have on node performance (eg block production latency), we have made these features run in the background
  - this works as expected, therefore pruning + compaction does not show to impact node performance and we recommend using it
- regarding key layouts:
  - our hypothesis was that the new "v2" layout should improve read/write performance on block & state store access
  - we obtained contradicting results on this hypothesis locally vs. production settings, concretely:
  - in local setting, enabling pruning with the old "v1" key layout _had_ a performance impact; enabling pruning with new layout _had no_ impact
  - in mainnet setting the observation was in reverse: using the new layout _introduced a ~10ms latency_
  - it is inconclusive if the new key layout "v2" is beneficial, so we will introduce this as an experimental feature
  - we expect to continue working with operator teams to gather data from production, ideally Injective and Osmosis
- pebbleDB: handles compaction without the need for Comet to force it, generally shows better performance with the new layout
```

Let's add a top-level summary of Q1 results here. Please double-check this is accurate Jasmina.

File: docs/references/storage/README.md [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+The two layouts are not interchange-able, once one is used, a node cannot switch to the other. The version to be used is set in the `config.toml` and defaults to `v1` - the current layout. Once the layout is set, it is written back to the database with `version` as the key. When a node boots up and loads the database initially, if this flag is set, it takes precedense over any configuration file. 
+
+The support for both layouts will allow users to benchmark their applications. If at any point we get clear indications that one layout is better than the other, we will gladly drop support for one of them and provide users with a way to migrate their databases gracefully. 
+
+
+## Pebble
+
+`PebbleDB` was recently added to `cometbft-db` by Notional labs and based on their benchmarks it was superior to goleveldDB. 
+
+We repeated our tests done in **1-node-local** using PebbleDB as the underlying database. While the difference in performance itself was slightly better, the most impressive difference is that PebbleDB seemed to handle compaction very well without a need to enforce it.
```
Comment (adizere) [inline docs/references/storage/README.md:242]:
```suggestion
We repeated our tests done in **1-node-local** using PebbleDB as the underlying database. While PebbleDB is slightly better in performance (tx/s), the most impressive difference is that PebbleDB seemed to handle compaction very well without a need to enforce it. 
```

**Comment by adizere** [comment]
Two more things:

- we should say which exact version of Comet did we use (except for Injective results) to get these results. It's `v1-alpha.1` or something from `main` ?
- we should add a mention to `docs/references/README.md` about this new file we're introducing. I did that in 20ec389 hope it's ok.

File: docs/references/storage/README.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+### Running CometBFT's e2e application in a mixed network of 6 nodes. 
+![e2e_storage_usage](img/e2e_storage_usage.png "Storage usage e2e")
+ 
+ The nodes doing pruning and compaction have a constant footprint compared to the other nodes. 
+ *validator03* and *validator05* prune without compaction. They have a smaller footprint than the 
+ nodes without pruning. The fact that the footprint of *validator05* has a lower footprint than 
+ *validator03* stems from the compaction logic of `goleveldb`. As the keys on *validator03* are sorted
+ by height, new data is simply appended without the need to reshuffle very old levels with old heights. 
+ On *validator05*, keys are sorted lexicographically leading to `goleveldb` *touching* more levels on insertions. By default, the conditions for triggering compaction are evaluated only when a file is touched. This is the reason why random key order leads to more frequent compaction. (This was also confirmed by findings done by our intern in Q3/Q4 2023 on goleveldb without Comet on top).
```
[the remark this answers is not in the document]
Comment (jmalicevic) [reply docs/references/storage/README.md:110]: That is a good point. We don't have a PR for Anton's work but I have a pdf with his presentation on the findings and links to his PDF results. A link to this doc is posted as a comment to https://github.com/cometbft/cometbft/issues/64 . We could add a bullet point with the doc in it as his findings should inform the work defined in 64.

**Comment by jmalicevic** [comment]
> Two more things:
> 
> * we should say which exact version of Comet did we use (except for Injective results) to get these results. It's `v1-alpha.1` or something from `main` ?

[Done](https://github.com/cometbft/cometbft/pull/2596/files/bb040ffb7ada1043c162a81e72d9e7dc8e8e2654..861edc0c02e72ec992e5bc577be712a909877105)

File: docs/references/storage/README.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+- we confirmed that moving pruning + compaction to background mitigates potential performance impact of this feature, by running tests with it on Injective mainnet
+  - this works as expected, therefore pruning + compaction does not show to impact node performance and we recommend using it
+- regarding key layouts:
+  - our hypothesis was that the new "v2" layout should improve read/write performance on block & state store access
+  - we obtained contradicting results on this hypothesis locally vs. production settings, concretely:
+  - in local setting, enabling pruning with the old "v1" key layout _had_ a performance impact; enabling pruning with new layout _had no_ impact
+  - in mainnet setting the observation was in reverse: using the new layout _introduced a ~10ms latency_
+  - it is inconclusive if the new key layout "v2" is beneficial, so we will introduce this as an experimental feature
+  - we expect to continue working with operator teams to gather data from production, ideally Injective and Osmosis
+- pebbleDB: handles compaction without the need for Comet to force it, generally shows better performance with the new layout
```
Comment (adizere) [inline docs/references/storage/README.md:56]: Let's add here what exists (pruning, forced compaction, background pruning+compaction) in which versions (v1/main/0.38/0.38-experimental) of CometBFT pls.

File: docs/references/storage/README.md
[the remark this answers is not in the document]
Comment (jmalicevic) [reply docs/references/storage/README.md:1]: Done

**Pull request merged**